### PR TITLE
Refactor to be more functional, refactor tests

### PR DESCRIPTION
**Pull request by durch**
This is quite a big PR, it however brings no new functionality it is purely refactoring and cleanup. First part of the refactoring does away with most of the GLOBAL variables in `background.js` and replaces them with functions. On top of that getter and setter functions are defined for handling mutating global maps. Configs are also refactored to a function.

Second (and bigger) part is the test refactor, in addition to switching all the tests to use new functions, all globals have been extracted to `jest/globals.js` which allows us to delete a lot of boilerplate from individual test files. All relevant getter and setter functions have been mocked and cleaned up.

In addition to the refactor `jest-each` is used to test any configs which have `id > 0`, in suits where such testing makes sense.

### CODE REVIEW

**Comment by durch** [comment]
@alxdavids 👍 on adding lint to tests, hope you like this one ;)

File: src/ext/background.js [hunk 1]
```diff
@@ -16,59 +16,66 @@
...
-let MAX_REDIRECT = ACTIVE_CONFIG["spending-restrictions"]["max-redirects"];
-let NEW_TABS = ACTIVE_CONFIG["spending-restrictions"]["new-tabs"];
-let BAD_NAV = ACTIVE_CONFIG["spending-restrictions"]["bad-navigation"];
-let BAD_TRANSITION = ACTIVE_CONFIG["spending-restrictions"]["bad-transition"];
-let VALID_REDIRECTS = ACTIVE_CONFIG["spending-restrictions"]["valid-redirects"];
-let VALID_TRANSITIONS = ACTIVE_CONFIG["spending-restrictions"]["valid-transitions"];
-let VAR_RESET = ACTIVE_CONFIG["var-reset"];
-let VAR_RESET_MS = ACTIVE_CONFIG["var-reset-ms"];
+let CONFIG_ID = 1;
+const getConfigId = () => CONFIG_ID;
```
Comment (alxdavids) [inline src/ext/background.js]: Should these go with the getters and setters below?

File: src/ext/background.js [hunk 2]
```diff
@@ -177,30 +211,30 @@ function processHeaders(details, url) {
...
     // If we have tokens to spend, cancel the request and pass execution over to the token handler.
     let attempted = false;
-    if (activated && !spentUrl[url.href]) {
+    if (activated && !getSpentUrl(url.href)) {
         const count = countStoredTokens();
-        if (DO_REDEEM) {
-            if (count > 0 && !url.host.includes(CHL_CAPTCHA_DOMAIN)) {
+        if (doRedeem()) {
+            if (count > 0 && !url.host.includes(chlCaptchaDomain())) {
                 attemptRedeem(url, details.tabId, target);
```
Comment (alxdavids) [inline src/ext/background.js]: Remove `target` from this invocation?

File: src/ext/browserUtils.js [hunk 3]
```diff
@@ -49,25 +48,22 @@ function attemptRedeem(url, respTabId, target) {
                 if (tabIds.includes(respTabId)) {
                     chrome.cookies.get({
                         "url": url.href,
-                        "name": CHL_CLEARANCE_COOKIE,
-                        "storeId": store.id,
-                    }, function(cookie) {
                         // Require an existing, non-expired cookie.
-                        if (cookie) {
+                        if(cookie) {
```
Comment (alxdavids) [inline src/ext/browserUtils.js]: Can we add this space back?

File: src/ext/browserUtils.js [hunk 4]
```diff
@@ -78,19 +74,19 @@ function attemptRedeem(url, respTabId, target) {
  * @param {Number} respTabId ID of the tab where the request originated
  * @param {Array<string>} target Array of possible targets for redemption
```
Comment (alxdavids) [inline src/ext/browserUtils.js]: Remove `target` from JSDoc

File: src/ext/browserUtils.js [hunk 5]
```diff
@@ -25,18 +25,17 @@
...
 "use strict";
-
-let CHECK_COOKIES = ACTIVE_CONFIG["cookies"]["check-cookies"];
+const CHECK_COOKIES = () => activeConfig()["cookies"]["check-cookies"];
 
 /**
  * Attempts to redeem a token if the series of checks passes
  * @param {URL} url URL of request
  * @param {Number} respTabId ID of the tab where the request originated
  * @param {Array<string>} target Array of possible targets for redemption
```
Comment (alxdavids) [inline src/ext/browserUtils.js]: Remove `target` from JSDoc

File: test/beforeRequest.test.js [hunk 6]
```diff
@@ -33,9 +30,7 @@ const parseSigString = workflow.__get__("parseSigString");
 const setConfig = workflow.__get__("setConfig");
 const getCachedCommitments = workflow.__get__("getCachedCommitments");
 const cacheCommitments = workflow.__get__("cacheCommitments");
-let localStorage;
-let details;
-let url;
+const PPConfigs = workflow.__get__("PPConfigs");
 const respGoodProof = `signatures=WyJCTGZOdGZ6eG92RXdwZk5LWVBvRkk3dHNLNk5rMjNNalluUklEVFhGdHEwYm9zbWJIN1l1bERYWHVrdVYrKytxZyttYU9UWEF4cXpGSHNkV3p2dEpmQU09IiwiQkVIa1BPT1p3UlIrT0dNKzJQTUJnRWVrdUhobVJpVUlJSGxiaGJqNkNSKzZ2blp3Sk1CTHlNbDR4aURuOVY4SUhQNFdENFRaTUJGQjR0cStXd0c5azdRPSIsIkJBKzM4NkZPNkNXODZJbGIxdzdEOUZWMytwRnN
```
Comment (alxdavids) [inline test/beforeRequest.test.js]: This test data might be better in jest/globals now?

File: test/curve.test.js [hunk 7]
```diff
@@ -218,12 +236,12 @@ describe("point compression/decompression", () => {
...
     const params = getCurveParams(curve);
     let eleFFp;
-    if (!t && t != 0) {
+    if (!t && t !== 0) {
         const byteLength = 32;
         const wordLength = byteLength / 4; // SJCL 4 bytes to a word
         const random = sjcl.random.randomWords(wordLength, 10);
         const rndBits = sjcl.codec.bytes.toBits(random);
-        eleFFp = h2Base(rndBits, curve, "p256_hashing");
+        eleFFp = h2Base(rndBits, "p256", "p256_hashing");
```
Comment (alxdavids) [inline test/curve.test.js]: Why has this been changed back to hardcoded?

**Comment by Lekensteyn** [comment]
Hi @durch, thanks for your PR. The current changes are quite large though, making review a bit harder. Would it be possible to split the changes in separate, meaningful commits with isolated changes?

For example, perform renaming changes in one (or more) change(s) without addressing other lint issues at the same time.

File: src/ext/browserUtils.js (hunk 3 above)
Comment (durch) [reply src/ext/browserUtils.js]: ESLint complains about `(space-before-function-paren)`, that seems wrong, I'll look into it

File: src/ext/background.js (hunk 1 above)
Comment (durch) [reply src/ext/background.js]: Unfortunately not, they're being used line below, so need to be here, or further up, I'll add a newline here for readability

**Comment by durch** [comment]
Hi @Lekensteyn, I'm not really sure how to split it up, switch to the functional approach, requires to also modify the tests, so thing pulls the other. I realise that the PR is quite big, it might make sense to review it one commit at a time. They are pretty well segmented, the rename is a separate commit, and different tests are grouped by commits. Open to other suggestions

**Comment by durch** [comment]
Circle CI seems to be having some issues

**Comment by alxdavids** [comment]
> Hi @Lekensteyn, I'm not really sure how to split it up, switch to the functional approach, requires to also modify the tests, so one thing pulls the other.
> 
> I realise that the PR is quite big, it might make sense to review it one commit at a time. They are pretty well segmented, the rename is a separate commit, and different tests are grouped by commits. Open to other suggestions

It looks like there are commits from the previous PR (https://github.com/privacypass/challenge-bypass-extension/pull/81) included in this one. Would it be possible to rebase this change on master and see if the number of commit reduces?

**Comment by durch** [comment]
That would definitely be possible, I seem to have messed up my merge orders here :/

**Comment by durch** [comment]
@alxdavids, @Lekensteyn, this was really a mess, I don't understand how I missed it, thank you for pointing it out and being polite about it :).

Would you like me to just close this and reopen another clean one, or are we fine to continue here? 

I'll do another pass through your feedback so there might be additional commits coming in

**Comment by alxdavids** [comment]
> @alxdavids, @Lekensteyn, this was really a mess, I don't understand how I missed it, thank you for pointing it out and being polite about it :).
> 
> Would you like me to just close this and reopen another clean one, or are we fine to continue here?
> 
> I'll do another pass through your feedback so there might be additional commits coming in

No problem, I think it might be better to open a new PR just so the history cleans up a bit and we can review it from fresh. 

By the way, for the circleci errors, I think it's because it is building on jessie so swapping line 6 of ` .circleci/config.yml` from `- image: circleci/node:8.12` to `- image: circleci/node:8.12-stretch` should fix it.

**Comment by alxdavids** [comment]
> > @alxdavids, @Lekensteyn, this was really a mess, I don't understand how I missed it, thank you for pointing it out and being polite about it :).
> > Would you like me to just close this and reopen another clean one, or are we fine to continue here?
> > I'll do another pass through your feedback so there might be additional commits coming in
> 
> No problem, I think it might be better to open a new PR just so the history cleans up a bit and we can review it from fresh.
> 
> By the way, for the circleci errors, I think it's because it is building on jessie so swapping line 6 of ` .circleci/config.yml` from `- image: circleci/node:8.12` to `- image: circleci/node:8.12-stretch` should fix it.

alternatively, rebase on top of the latest master

**Comment by durch** [comment]
@alxdavids thank you :), I' will go with a new PR and a rebase. It will be cleaner that way.

**Pull request closed without merging**